Module commands to have ACL categories.

roshkhatri opened the pull request:
Related to issues #11516 #11083 

This allows modules to register commands to existing ACL categories and blocks the creation of [sub]commands, datatypes and registering the configs outside of the OnLoad function.

For allowing modules to register commands to existing ACL categories,
This PR implements a new API `int RM_SetCommandACLCategories()` which takes a pointer to a `RedisModuleCommand` and a C string `aclflags` containing the set of space separated ACL categories.
Example, `'write slow'` marks the command as part of the write and slow ACL categories.

The C string `aclflags` is tokenized by implementing a helper function `categoryFlagsFromString()`. Theses tokens are matched and the corresponding ACL categories flags are set by a helper function `matchAclCategoriesFlags`. The helper function `categoryFlagsFromString()` returns the corresponding `categories_flags` or returns `-1` if some token not processed correctly.

If the module contains commands which are registered to existing ACL categories, the number of [sub]commands are tracked by `num_commands_with_acl_categories` in `struct RedisModule`. Further, the allowed command bit-map of the existing users are recomputed from the `command_rules` list, by implementing a function called `ACLRecomputeCommandBitsFromCommandRulesAllUsers()` for the existing users to have access to the module commands on runtime.


## Breaking change

This change requires that registering commands and subcommands only occur during a modules "OnLoad" function, in order to allow efficient recompilation of ACL bits. We also chose to block registering configs and types, since we believe it's only valid for those to be created during onLoad. We check for this `onload` flag in `struct RedisModule` to check if the call is made from the `OnLoad` function.

madolson commented on src/module.c:
```diff
@@ -11674,6 +11709,8 @@ int moduleUnload(sds name) {
     module->name = NULL; /* The name was already freed by dictDelete(). */
     moduleFreeModuleStructure(module);
 
+    /* Recompute Commmand Bits for all users once the modules has been completely unloaded */
```
```suggestion
    /* Recompute command bits for all users once the modules has been completely unloaded */
```

roshkhatri replied:
Made the necessary changes.

madolson commented on src/acl.c:
```diff
@@ -638,6 +638,43 @@ void ACLSetSelectorCommandBitsForCategory(dict *commands, aclSelector *selector,
     dictReleaseIterator(di);
 }
 
+/* This funtion is responsible for recomputing the command bits for all selectors of the existing users.
```
```suggestion
/* This function is responsible for recomputing the command bits for all selectors of the existing users.
```

roshkhatri replied:
Made the necessary changes.

madolson commented on src/acl.c:
```diff
@@ -638,6 +638,43 @@ void ACLSetSelectorCommandBitsForCategory(dict *commands, aclSelector *selector,
+    raxIterator ri;
+    raxStart(&ri,Users);
+    raxSeek(&ri,"^",NULL,0);
+    while(raxNext(&ri)){
```
```suggestion
    while(raxNext(&ri)) {
```

madolson commented on src/acl.c:
```diff
@@ -638,6 +638,43 @@ void ACLSetSelectorCommandBitsForCategory(dict *commands, aclSelector *selector,
+        listIter li;
+        listNode *ln;
+        listRewind(u->selectors,&li);
+        while((ln = listNext(&li))){
```
```suggestion
        while((ln = listNext(&li))) {
```

madolson commented on src/module.c:
```diff
@@ -1186,6 +1213,10 @@ int RM_CreateCommand(RedisModuleCtx *ctx, const char *name, RedisModuleCmdFunc c
     cp->rediscmd->id = ACLGetCommandID(declared_name); /* ID used for ACL. */
+    cp->rediscmd->acl_categories = acl_categories; /* ACL categories flags for module command */
+    /* If commands have acl_categories, recompute the command bits for the existing 
+     *  users to have access to these commands */  
```
```suggestion
     * users to have access to these commands */  
```

madolson reviewed (approved):
nitpicks

madolson commented:
@redis/core-team Hey team. This is a new module API that allows module commands to register themselves to the built-in ACL categories.

oranagra commented on src/acl.c:
```diff
@@ -627,7 +627,7 @@ void ACLSetSelectorCommandBitsForCategory(dict *commands, aclSelector *selector,
     while ((de = dictNext(di)) != NULL) {
         struct redisCommand *cmd = dictGetVal(de);
-        if (cmd->flags & CMD_MODULE) continue; /* Ignore modules commands. */
+        if (cmd->flags & CMD_MODULE && !(cmd->acl_categories)) continue; /* Ignore modules commands unless they have acl category. */
```
why not just remove that line?
the code below will behave correctly when cmd->acl_categories is 0

oranagra commented on src/acl.c:
```diff
@@ -638,6 +638,43 @@ void ACLSetSelectorCommandBitsForCategory(dict *commands, aclSelector *selector,
+            int argc = 0;
+            sds *argv = sdssplitargs(selector->command_rules, &argc);
+            serverAssert(argv != NULL);
+            if (ACLSelectorCanExecuteFutureCommands(selector)) {
+                int res = ACLSetSelector(selector,"+@all",-1);
+                serverAssert(res == C_OK);
+            } else {
+                int res = ACLSetSelector(selector,"-@all",-1);
+                serverAssert(res == C_OK);
+            }
```
let's add a comment for this block saying that it's done to clean the slate.
(assuming i'm right)

oranagra replied:
btw, why not use `selector->flags & SELECTOR_FLAG_ALLCOMMANDS` instead of ACLSelectorCanExecuteFutureCommands? or am i missing something?

madolson replied:
> btw, why not use selector->flags & SELECTOR_FLAG_ALLCOMMANDS instead of ACLSelectorCanExecuteFutureCommands? or am i missing something?

They do the same thing. I assume part of this is parity with ACLDescribeSelectorCommandRules which does the same thing as part of the "sanity" check code. We could probably simplify the code.

roshkhatri replied:
> let's add a comment for this block saying that it's done to clean the slate. (assuming i'm right)

yes!

> > btw, why not use selector->flags & SELECTOR_FLAG_ALLCOMMANDS instead of ACLSelectorCanExecuteFutureCommands? or am i missing something?
> 
> They do the same thing. I assume part of this is parity with ACLDescribeSelectorCommandRules which does the same thing as part of the "sanity" check code. We could probably simplify the code.

I will look into it!

madolson replied:
> I will look into it!

Just going to say this is not a high priority, let's consider refactoring this later.

oranagra commented on src/module.c:
```diff
@@ -1032,11 +1032,29 @@ int isCommandNameValid(const char *name) {
+        uint64_t this_flag = ACLGetCommandCategoryFlagByName(flag + 1);
+        if (this_flag) {
+            *acl_categories |= (int64_t) this_flag;
+            serverLog(LL_NOTICE,"Recognized acl category flag %s on module load", flag);
```
why do we need / want that log?
maybe the log level should be reduced, or if we really want to have some warning present when modules use this feature, maybe a more official introspection feature is needed (in INFO or MODULE command)

oranagra commented on src/module.c:
```diff
@@ -1032,11 +1032,29 @@ int isCommandNameValid(const char *name) {
  *
  * It returns the set of flags, or -1 if unknown flags are found. */
-int64_t commandFlagsFromString(char *s) {
+int64_t commandFlagsFromString(char *s, int64_t *acl_categories) {
```
maybe resetting `acl_categories` to 0 should be the responsibility of this function and not the caller (being an output parameter).

oranagra commented on src/module.c:
```diff
@@ -1186,6 +1213,10 @@ int RM_CreateCommand(RedisModuleCtx *ctx, const char *name, RedisModuleCmdFunc c
+    cp->rediscmd->acl_categories = acl_categories; /* ACL categories flags for module command */
+    /* If commands have acl_categories, recompute the command bits for the existing 
+     * users to have access to these commands */  
+    if (acl_categories) ACLRecomputeCommandBitsFromCommandRulesAllUsers(); 
```
maybe doing this per registered command is a bit harsh.
i.e. module that registers 10 commands, will re-run though all users for each command.
considering a module is only supposed to register commands when loaded, maybe we can raise some flag, and do this when the loading completed and all commands were registered.

roshkhatri replied:
Yes, that sounds good!

oranagra replied:
@MeirShpilraien @madolson i suppose that technically, till now, we allowed modules to register commands even after the OnLoad. am i right? do you see any reason to keep this possibility?

MeirShpilraien replied:
@oranagra I do not see why we should keep (none of our modules does it).

madolson replied:
@oranagra I'm also fine dropping that.

oranagra replied:
so shall we add some check in RM_CreateCommand and maybe also RM_CreateDataType and RM_Register*Config, RM_LoadConfigs that verifies we're inside the OnLoad?

madolson replied:
Those makes sense. The only other one I would consider is RedisModule_RegisterCommandFilter, which I assume shouldn't really be messed with at runtime either but maybe there are some use cases. Also presumably RM_CreateSubCommand.

oranagra replied:
for sure sub-commands goes together with commands.
i'm not sure about the filter, we don't really care do we?

oranagra replied:
please describe this change (blocked APIs after init) in the PR top comment.
as well as anything else worth noting for whoever won't read the code.

oranagra commented on tests/unit/moduleapi/aclcheck.tcl:
```diff
@@ -92,6 +92,30 @@ start_server {tags {"modules acl"}} {
+        assert_equal [r module unload aclcheck] OK
+        r acl SETUSER j3 on >password -@all +@WRITE
+        assert_equal [r module load $testmodule] OK
+        assert_equal [r acl DRYRUN j3 aclcheck.module.command.aclcategories.write] OK
```
maybe add a test that verifies that when loading module after user was created, it doesn't accidentally grant the user access that it shouldn't have had.

roshkhatri replied:
Do you mean, similar to the test next to this one, but where users are already present and modules are loaded later. will do that

oranagra replied:
Exactly

oranagra reviewed (commented):
i'd like @guybe7 to take a quick look

roshkhatri commented:
Hi team, I have pushed the code with the necessary changes. Please take a look and review them.
Thanks!

oranagra commented on src/module.c:
```diff
@@ -1214,9 +1219,10 @@ int RM_CreateCommand(RedisModuleCtx *ctx, const char *name, RedisModuleCmdFunc c
+    /* If commands have acl_categories, and if 'module_contains_aclcategories_flag' is already not set to 1,
+     * then set it to 1, to recompute the command bits for the existing users to have access to these
+     * commands once the module is loaded. */ 
+    if (acl_categories && !module_contains_aclcategories_flag) module_contains_aclcategories_flag = 1; 
```
this seems overly complicated (the check for `!module_contains_aclcategories_flag`), also, maybe it's nicer to do `++` instead of `= 1`

guybe7 replied:
i think this should be inside `struct RedisModule` and represent the number of [sub]commands that have ACl categories (i.e. also rename it to "num_commands_with_acl_categories"

oranagra commented on src/module.c:
```diff
@@ -11608,6 +11615,13 @@ int moduleLoad(const char *path, void **module_argv, int module_argc, int is_loa
+    ACLRecomputeCommandBitsFromCommandRulesAllUsers();
+    if(module_contains_aclcategories_flag) {
+        ACLRecomputeCommandBitsFromCommandRulesAllUsers();
+        module_contains_aclcategories_flag = 0;
```
i think it's nicer to unconditionally set it to 0 at the beginning of this function rather than re-set it to 0 when modified.

oranagra commented:
we need to consider how this impacts users on upgrade.
i.e. a command is added to a category implicitly exposes it (without user's intention), or blocks it (breaking an app).

i suppose this potential for breakage is not about upgrading redis (to include this API), but rather about upgrading the module. and not necessarily when the module starts using this API, but any time it changes the categories (i.e. even 5 years from now, when a module upgrade adds a command to a category).
and in fact it's arguably no different than cases were redis [changes](https://github.com/redis/redis/pull/9208) categories.

i guess it ultimately depends on the module developers, not redis. and like many other breaking changes, we just have a choice between keeping something bad (inconsistent, confusing or hard to use) forever, or fixing it and accepting the uncomfortable transition. still, i'd like to hear opinions and suggestions on how we can mitigate this.

guybe7 commented on src/module.c:
```diff
@@ -1062,7 +1085,11 @@ int64_t commandFlagsFromString(char *s) {
         else if (!strcasecmp(t,"no-mandatory-keys")) flags |= CMD_NO_MANDATORY_KEYS;
         else if (!strcasecmp(t,"allow-busy")) flags |= CMD_ALLOW_BUSY;
-        else break;
+        else if (!matchAclCategoriesFlags(t, acl_categories));
+        else {
+            serverLog(LL_WARNING,"Unrecognized command flag %s on module load", t);
+            break;
+        }
```
`matchAclCategoriesFlags` returns REDISMODULE_OK/REDISMODULE_ERR so `!` should not be used here (the reason it worked is that OK is 0 and ERR isn't)
regardless, i find this code structure a bit confusing, why not:
```
else {
    if (matchAclCategoriesFlags(t, acl_categories) == REDISMODULE_ERR) {
        serverLog(LL_WARNING,"Unrecognized command flag %s on module load", t);
        break;
   }
}
```

madolson replied:
I'm fine with the new structure, but I don't really think matchAclCategoriesFlags should return REDISMODULE_ERR or REDISMODULE_OK now that I think about it, since those are more typically used to module error codes. Since this is an internal function, it should probably directly return 1 and 0.

oranagra commented:
this PR was conceptually approved in a core-team meeting.
can be merged after the last batch of comments are addressed
we are ok with blocking the registration of command, configs, and data type outside the onload (must be mentioned in the top comment as a breaking change).

one other concern that was raised is if the API would be convenient. considering that all modules are gonna want to use this, but they must remain compatible with older versions of redis, they need to do some filtering of the categories from the flag string they pass.
the alternative is to add this feature to either a new API or the existing `RM_SetCommandInfo`, so the module can conditionally add this info without filtering of the hard coded flags string.
we would like to explore these alternative (implement each in some branch) and see which one feels better.

@roshkhatri can you handle all of that in the coming couple of weeks?

roshkhatri commented:
Thanks, that sounds good, yes I can handle these changes in couple of weeks. Will keep you posted on the PR. Will implement the alternative and put up another PR to see which one makes more sense.

roshkhatri commented:
Hi Team, I have updated this PR and have addressed the comments. I have also updated the PR with the implementation of blocking the registration of command, registering the configs, and creating the data type outside the OnLoad function. Please review the changes and let me know if there are any questions.

madolson commented:
For the conversation about RedisModule_SetCommandInfo(). I feel like there are three options:
1. We leave the API as is. Module developers will need to implement a check for Redis version and determine flags as necessary.
2. We extend RedisModule_SetCommandInfo(). We will need a new version, and module developers will still need to check for Redis version for setting the fields.
3. We implement a third API. RedisModule_SetCommandCategories(), which a module developer will need to check the version to know if they are allowed to call. 

\1 and 2\ seem equally bad, as we basically are required to have a Redis version check for the API. 3\ seems theoretically better, as we could check if that command exists, and only call it if it does. Implementation wise they are all pretty similar now, as we just need to go and set all the bits and flags and call "recompute" like we already do.

In the fullness of time, I still think 1 is the clearest and most sensible approach. I would still be inclined to suggest we merge what is currently implemented. @yossigo Have you thought much more about this.

madolson commented on src/module.c:
```diff
@@ -1285,9 +1317,13 @@ RedisModuleCommand *RM_GetCommand(RedisModuleCtx *ctx, const char *name) {
  * * `parent` is already a subcommand (we do not allow more than one level of command nesting)
  * * `parent` is a command with an implementation (RedisModuleCmdFunc) (A parent command should be a pure container of subcommands)
  * * `parent` already has a subcommand called `name`
+ * * Creating a subcommand is called outside of the RedisModule_OnLoad.
```
```suggestion
 * * Creating a subcommand is called outside of RedisModule_OnLoad.
```

madolson commented on src/module.c:
```diff
@@ -6484,7 +6524,8 @@ robj *moduleTypeDupOrReply(client *c, robj *fromkey, robj *tokey, int todb, robj
  * happens to be pretty lame as well.
  *
- * If there is already a module registering a type with the same name,
+ * If RedisModule_CreateDataType() is called outside of the RedisModule_OnLoad() function,
+ * and if there is already a module registering a type with the same name,
  * and if the module name or encver is invalid, NULL is returned.
```
```suggestion
 * If RedisModule_CreateDataType() is called outside of RedisModule_OnLoad() function,
 * there is already a module registering a type with the same name,
 * or if the module name or encver is invalid, NULL is returned.
```

madolson commented on src/module.c:
```diff
@@ -11573,7 +11616,13 @@ int moduleLoad(const char *path, void **module_argv, int module_argc, int is_loa
 
+    /* If module commands have ACL categories, recompute command bits 
+     * for all existing users once the modules has been registered. */
+    if(ctx.module->num_commands_with_acl_categories) {
```
```suggestion
    if (ctx.module->num_commands_with_acl_categories) {
```

madolson commented on src/module.c:
```diff
@@ -12203,10 +12259,11 @@ int RM_RegisterNumericConfig(RedisModuleCtx *ctx, const char *name, long long de
 
 /* Applies all pending configurations on the module load. This should be called
  * after all of the configurations have been registered for the module inside of RedisModule_OnLoad.
+ * This will return REDISMODULE_ERR if it is called outside the RedisModule_OnLoad.
```
```suggestion
 * This will return REDISMODULE_ERR if it is called outside RedisModule_OnLoad.
```

oranagra commented on src/module.c:
```diff
@@ -1032,11 +1032,29 @@ int isCommandNameValid(const char *name) {
+ *
+ * Returns '1' if acl category flag is recognized or
+ * returns '0' if not recognized  */
+int matchAclCategoriesFlags(char *flag, int64_t *acl_categories){
```
```suggestion
int matchAclCategoriesFlags(char *flag, int64_t *acl_categories) {
```

oranagra commented on src/module.c:
```diff
@@ -1061,8 +1079,11 @@ int64_t commandFlagsFromString(char *s) {
-        else if (!strcasecmp(t,"allow-busy")) flags |= CMD_ALLOW_BUSY;
-        else break;
+        else if (!strcasecmp(t,"allow-busy")) flags |= CMD_ALLOW_BUSY; 
+        else if (!matchAclCategoriesFlags(t, acl_categories)){
```
```suggestion
        else if (!strcasecmp(t,"allow-busy")) flags |= CMD_ALLOW_BUSY;
        else if (!matchAclCategoriesFlags(t, acl_categories)) {
```

oranagra commented on src/module.c:
```diff
@@ -2134,6 +2130,8 @@ void RM_SetModuleAttribs(RedisModuleCtx *ctx, const char *name, int ver, int api
     module->defrag_cb = 0;
     module->loadmod = NULL;
+    module->num_commands_with_acl_categories = 0;
+    module->onload = 1;
```
maybe it'll be nicer to set `onload` in `moduleLoad` rather than here?

roshkhatri replied:
Here we set the `onload` flag in `RM_InIt()` from `(*onload)(void *, void **, int)` in `moduleLoad` function, as far as I know the module APIs are called from this `onlaod` function.

oranagra commented on tests/modules/aclcheck.c:
```diff
@@ -190,6 +190,23 @@ int module_test_acl_category(RedisModuleCtx *ctx, RedisModuleString **argv, int
+    error |= (result == REDISMODULE_ERR);
+    RedisModuleCommand *parent = RedisModule_GetCommand(ctx,"block.commands.outside.onload");
+    result = RedisModule_CreateSubcommand(parent,"subcommand.that.should.fail",module_test_acl_category,"",0,0,0);
+    error |= (result == REDISMODULE_ERR);
```
i suppose we want to fail if either of them succeeds.
if i read the code correctly, the "NOPERM" which we expect, will happen if at least one of them fails.

p.s. i'm not sure we should go to that level of detail in the testing, i would be willing to test just one of them, or none (not all 8 APIs)

oranagra commented on tests/modules/moduleconfigs.c:
```diff
@@ -103,6 +103,35 @@ int longlongApplyFunc(RedisModuleCtx *ctx, void *privdata, RedisModuleString **e
+    error |= (result == REDISMODULE_ERR);
+
+    result = RedisModule_RegisterStringConfig(ctx, "string", "secret password", REDISMODULE_CONFIG_DEFAULT, getStringConfigCommand, setStringConfigCommand, NULL, NULL);
+    error |= (result == REDISMODULE_ERR);
```
same issue here.
we probably want to count successes, not errors.

oranagra commented:
> \1 and 2\ seem equally bad, as we basically are required to have a Redis version check for the API. 3\ seems theoretically better, as we could check if that command exists, and only call it if it does. Implementation wise they are all pretty similar now, as we just need to go and set all the bits and flags and call "recompute" like we already do.

IIRC the concern wasn't about detecting the redis version, but rather about the string manipulation the module would need to do (either concatenate strings, or filter them), which would make the boilerplate code of registering commands rather ugly.
so if that's the case, and we take the categories to another string (either SetCommandInfo or SetCommandCategories), it resolves that problem.
i agree that in the fullness of time this concern is gonna be gone, but in the short term (next 4 years), all modules that wanna use it are gonna suffer from this complication.

am i missing something?

madolson commented:
Core meeting update. We decided that we should probably implement the new API, RM_SetcommandCategories(), since that was probably the most module developer friendly thing to do. Other than that, the new APIs are approved.

roshkhatri commented:
Hi Team, I have pushed the changes to this PR and implemented the new API, RM_SetcommandCategories(). Please review the changes and let me know if there are any questions. @oranagra  @madolson

madolson commented on src/module.c:
```diff
@@ -1061,7 +1061,7 @@ int64_t commandFlagsFromString(char *s) {
         else if (!strcasecmp(t,"no-cluster")) flags |= CMD_MODULE_NO_CLUSTER;
         else if (!strcasecmp(t,"no-mandatory-keys")) flags |= CMD_NO_MANDATORY_KEYS;
-        else if (!strcasecmp(t,"allow-busy")) flags |= CMD_ALLOW_BUSY;
+        else if (!strcasecmp(t,"allow-busy")) flags |= CMD_ALLOW_BUSY; 
```
```suggestion
        else if (!strcasecmp(t,"allow-busy")) flags |= CMD_ALLOW_BUSY;
```

madolson commented on src/module.c:
```diff
@@ -1341,6 +1348,56 @@ moduleCmdArgAt(const RedisModuleCommandInfoVersion *version,
+    for (j = 0; j < count; j++) {
+        char *t = tokens[j];
+        if (!matchAclCategoriesFlags(t, &acl_categories_flags)) {
+                serverLog(LL_WARNING,"Unrecognized categories flag %s on module load", t);
+                break;
```
```suggestion
            serverLog(LL_WARNING,"Unrecognized categories flag %s on module load", t);
            break;
```

madolson commented on src/module.c:
```diff
@@ -1341,6 +1348,56 @@ moduleCmdArgAt(const RedisModuleCommandInfoVersion *version,
+ * the module commands by acl categories.
+ * 
+ * It returns the set of acl flags, or -1 if unknown flags are found. */
+int64_t categoriesFlagsFromString(char *aclflags) {
```
```suggestion
int64_t categoryFlagsFromString(char *aclflags) {
```

madolson commented on src/module.c:
```diff
@@ -1341,6 +1348,56 @@ moduleCmdArgAt(const RedisModuleCommandInfoVersion *version,
     return (RedisModuleCommandArg *)((char *)(args) + offset);
 }
 
+/* Helper for categoriesFlagsFromString(). Turns a string representing command
```
```suggestion
/* Helper for categoryFlagsFromString(). Turns a string representing command
```

madolson commented on src/module.c:
```diff
@@ -1341,6 +1348,56 @@ moduleCmdArgAt(const RedisModuleCommandInfoVersion *version,
+    return 0; /* Unrecognized */
+}
+
+/* Helper for RM_SetCommandCategories(). Turns a string representing command
```
```suggestion
/* Helper for RM_SetCommandCategories(). Turns a string representing acl category
```

madolson commented on src/module.c:
```diff
@@ -1341,6 +1348,56 @@ moduleCmdArgAt(const RedisModuleCommandInfoVersion *version,
+ * commands and subcommands. The set of ACL categories should be passed as
+ * a space separated C string 'aclflags'.
+ * 
+ * Example, 'write' marks the command as part of the write ACL category. */
```
```suggestion
 * Example, 'write slow' marks the command as part of the write and slow ACL categories. */
```

madolson reviewed (commented):
Some minor documentation updates, will just commit them.

madolson commented on src/module.c:
```diff
@@ -1341,6 +1348,56 @@ moduleCmdArgAt(const RedisModuleCommandInfoVersion *version,
+ * Example, 'write slow' marks the command as part of the write and slow ACL categories. */
+int RM_SetCommandCategories(RedisModuleCommand *command, const char *aclflags) {
+    if (!command || !command->module || !command->module->onload) return REDISMODULE_ERR;
+    int64_t categories_flags = aclflags ? categoriesFlagsFromString((char*)aclflags) : 0;
```
```suggestion
    int64_t categories_flags = aclflags ? categoryFlagsFromString((char*)aclflags) : 0;
```

roshkhatri commented:
I have also updated the top comment with the updated implementation.

madolson commented on src/module.c:
```diff
@@ -1341,6 +1348,56 @@ moduleCmdArgAt(const RedisModuleCommandInfoVersion *version,
+ *
+ * Returns '1' if acl category flag is recognized or
+ * returns '0' if not recognized  */
+int matchAclCategoriesFlags(char *flag, int64_t *acl_categories_flags) {
```
```suggestion
int matchAclCategoryFlag(char *flag, int64_t *acl_categories_flags) {
```

madolson commented on src/module.c:
```diff
@@ -1341,6 +1348,56 @@ moduleCmdArgAt(const RedisModuleCommandInfoVersion *version,
+    sds *tokens = sdssplitlen(aclflags,strlen(aclflags)," ",1,&count);
+    for (j = 0; j < count; j++) {
+        char *t = tokens[j];
+        if (!matchAclCategoriesFlags(t, &acl_categories_flags)) {
```
```suggestion
        if (!matchAclCategoryFlag(t, &acl_categories_flags)) {
```

madolson commented on src/module.c:
```diff
@@ -1341,6 +1348,56 @@ moduleCmdArgAt(const RedisModuleCommandInfoVersion *version,
     return (RedisModuleCommandArg *)((char *)(args) + offset);
 }
 
+/* Helper for categoryFlagsFromString(). Turns a string representing command
+ * flags into the ACL Category Flags.
```
```suggestion
/* Helper for categoryFlagsFromString(). Attempts to find an acl flag representing the provided flag string
 * and adds that flag to acl_categories_flags if a match is found.
```

madolson reviewed (approved):
Tweaked top level comment, and some other minor wording suggestions which I will apply.

madolson commented:
@oranagra @yossigo All updates, PTAL and validate the new name looks good at the very least.

oranagra commented on tests/modules/aclcheck.c:
```diff
@@ -193,13 +193,15 @@ int module_test_acl_category(RedisModuleCtx *ctx, RedisModuleString **argv, int
+    if (response_ok) {
         RedisModule_ReplyWithError(ctx, "NOPERM");
     } else {
         RedisModule_ReplyWithSimpleString(ctx, "OK");
```
it looks a bit suspicious that we return "OK" when things failed.
maybe a comment saying that it validates that it's not possible to register these after loading, thus return an error if they succeed.

oranagra replied:
or we can flip it and return back to using `assert_error {NOPERM}`.
i.e. my complaint was that the `|=` tested that at least one failed, rather than at least one succeeded.

roshkhatri replied:
asserting the error makes more sense along with changing the name of the test function from `block.commands.outside.onload` to `create.commands.outside.onload`

roshkhatri replied:
Yes, I will add a comment mentioning that and changing the error from `NOPERM` to `UNEXPECTEDOK` so it will make it more clear.

oranagra commented on src/module.c:
```diff
@@ -1379,6 +1348,56 @@ moduleCmdArgAt(const RedisModuleCommandInfoVersion *version,
+    return acl_categories_flags;
+}
+
+/* RedisModule_SetCommandCategories can be used to set ACL categories to module
```
I'm not sure about the name, i think it's missing a reference to the term "ACL"
arguably `RedisModule_SetCommandAclCategories` isn't that much longer...

madolson replied:
Sure, however I made it RedisModule_SetCommandACLCategories to make it consistent casing with other places we use ACL in module names.

roshkhatri replied:
yeah that makes sense, will change the name to `RedisModule_SetCommandAclCategories`

oranagra commented on tests/modules/aclcheck.c:
```diff
@@ -193,6 +219,30 @@ int RedisModule_OnLoad(RedisModuleCtx *ctx, RedisModuleString **argv, int argc)
+        return REDISMODULE_ERR;
+    RedisModuleCommand *aclcategories_write = RedisModule_GetCommand(ctx,"aclcheck.module.command.aclcategories.write");
+
+    if (RedisModule_SetCommandCategories(aclcategories_write, "write") == REDISMODULE_ERR)
```
aren't we missing a check that validates this API can't be used after loading?
after all, that was the main reason to add all these checks.

roshkhatri replied:
yeah, I will add a test to check this as well.

oranagra commented on src/module.c:
```diff
@@ -1076,6 +1076,7 @@ RedisModuleCommand *moduleCreateCommandProxy(struct RedisModule *module, sds dec
  * convention.
  *
  * The function returns REDISMODULE_ERR in these cases:
+ * - If creation of module command is called outside the RedisModule_OnLoad.
```
we're missing a similar comment in the new API.
